### ISSUE TITLE
Using FTPS after upgrading from 1.3.8a to 1.3.8b leads to crash

### ISSUE THREAD

**Issue by conathan** [proftpd/proftpd#1770]
### What I Did

Backporting proftpd 1.3.8b from AlmaLinux 9 to AlmaLinux 8
* When using FTPS(Port 21), crashes with a Stack Trace
* No issues with unencrypted FTP, or SFTP

Backporting proftpd 1.3.8a from AlmaLinux 9 to AlmaLinux 8 works with no known issues
* From https://src.fedoraproject.org/rpms/proftpd/commits/epel9
* They added support for libsodium.  I attempted the same test with that disabled disabled when I got the below crash, but no change
```
-----BEGIN STACK TRACE-----
[0] /lib64/libcrypto.so.1.1(X509_PUBKEY_get0_param+0x2b) [0x7f8291f87c0b]
[1] /lib64/libcrypto.so.1.1(X509_PUBKEY_get0_param+0x2b) [0x7f8291f87c0b]
[2] proftpd: 70.66.32.130:54188: AUTH TLS(+0xd23af) [0x558b67cfa3af]
[3] proftpd: 70.66.32.130:54188: AUTH TLS(+0xd339c) [0x558b67cfb39c]
[4] proftpd: 70.66.32.130:54188: AUTH TLS(+0xd99fe) [0x558b67d019fe]
[5] proftpd: 70.66.32.130:54188: AUTH TLS(pr_module_call+0x48) [0x558b67c7ffe8]
[6] proftpd: 70.66.32.130:54188: AUTH TLS(+0x29b92) [0x558b67c51b92]
[7] proftpd: 70.66.32.130:54188: AUTH TLS(pr_cmd_dispatch_phase+0x3e1) [0x558b67c547f1]
[8] proftpd: 70.66.32.130:54188: AUTH TLS(+0x2cac5) [0x558b67c54ac5]
[9] proftpd: 70.66.32.130:54188: AUTH TLS(+0x2abb4) [0x558b67c52bb4]
[10] proftpd: 70.66.32.130:54188: AUTH TLS(+0x2b516) [0x558b67c53516]
[11] proftpd: 70.66.32.130:54188: AUTH TLS(main+0x5e3) [0x558b67c511e3]
[12] /lib64/libc.so.6(__libc_start_main+0xe5) [0x7f829119dd85]
[13] proftpd: 70.66.32.130:54188: AUTH TLS(_start+0x2e) [0x558b67c5185e]
 -----END STACK TRACE-----
```

### What I Expected/Wanted

Successful login via ftps on port 21

### ProFTPD Version and Configuration
```
Compile-time Settings:
  Version: 1.3.8b (maint)
  Platform: LINUX [Linux 4.18.0-477.27.1.el8_8.x86_64 x86_64]
  OS/Release:
    NAME="AlmaLinux"
    VERSION="8.9 (Midnight Oncilla)"
    ID="almalinux"
    ID_LIKE="rhel centos fedora"
    VERSION_ID="8.9"
    PLATFORM_ID="platform:el8"
    PRETTY_NAME="AlmaLinux 8.9 (Midnight Oncilla)"
    CPE_NAME="cpe:/o:almalinux:almalinux:8::baseos"
    
    ALMALINUX_MANTISBT_PROJECT="AlmaLinux-8"
    ALMALINUX_MANTISBT_PROJECT_VERSION="8.9"
    REDHAT_SUPPORT_PRODUCT="AlmaLinux"
    REDHAT_SUPPORT_PRODUCT_VERSION="8.9"
  Built: Mon Jan 8 2024 13:16:53 PST
  Built With:
    configure  '--build=x86_64-redhat-linux-gnu' '--host=x86_64-redhat-linux-gnu' '--program-prefix=' '--disable-dependency-tracking' '--prefix=/usr' '--exec-prefix=/usr' '--bindir=/usr/bin' '--sbindir=/usr/sbin' '--sysconfdir=/etc' '--datadir=/usr/share' '--includedir=/usr/include' '--libdir=/usr/lib64' '--libexecdir=/usr/libexec' '--localstatedir=/var' '--sharedstatedir=/var/lib' '--mandir=/usr/share/man' '--infodir=/usr/share/info' '--libexecdir=/usr/libexec/proftpd' '--localstatedir=/run/proftpd' '--disable-strip' '--enable-ctrls' '--enable-dso' '--enable-facl' '--enable-ipv6' '--enable-memcache' '--enable-nls' '--enable-openssl' '--disable-pcre' '--enable-pcre2' '--disable-redis' '--enable-shadow' '--enable-tests=nonetwork' '--with-libraries=/usr/lib64/mariadb' '--with-includes=/usr/include/mysql' '--with-modules=mod_readme:mod_auth_pam:mod_tls' '--with-shared=mod_sql:mod_sql_passwd:mod_sql_mysql:mod_sql_postgres:mod_sql_sqlite:mod_quotatab:mod_quotatab_file:mod_quotatab_ldap:mod_quotatab_radius:mod_quotatab_sql:mod_ldap:mod_ban:mod_ctrls_admin:mod_facl:mod_load:mod_vroot:mod_radius:mod_ratio:mod_rewrite:mod_site_misc:mod_exec:mod_shaper:mod_wrap2:mod_wrap2_file:mod_wrap2_sql:mod_copy:mod_deflate:mod_ifversion:mod_qos:mod_sftp:mod_sftp_pam:mod_sftp_sql:mod_tls_shmcache:mod_tls_memcache:mod_unique_id:mod_ifsession' 'build_alias=x86_64-redhat-linux-gnu' 'host_alias=x86_64-redhat-linux-gnu' 'CFLAGS=-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection' 'LDFLAGS=-Wl,-z,relro  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld' 'CXXFLAGS=-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection'

  CFLAGS: -g2 -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -Wall -fno-omit-frame-pointer -fno-strict-aliasing -Werror=implicit-function-declaration
  LDFLAGS: -Wl,-L$(top_srcdir)/lib,-L$(top_builddir)/lib -Wl,-z,relro  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -rdynamic -L/usr/lib64/mariadb -L/usr/lib64/ -L/usr/lib64
  LIBS: -lacl  -lpcre2-posix -lpcre2-8 -lssl -lcrypto -lsodium -lcap  -lssl -lcrypto  -lpam -lattr -lidn2 -lresolv -lresolv -lcrypt -ldl -lmemcachedutil -lmemcached  -pthread

  Files:
    Configuration File:
      /etc/proftpd.conf
    Pid File:
      /run/proftpd/proftpd.pid
    Scoreboard File:
      /run/proftpd/proftpd.scoreboard
    Header Directory:
      /usr/include/proftpd
    Shared Module Directory:
      /usr/libexec/proftpd

  Info:
    + Max supported UID: 4294967295
    + Max supported GID: 4294967295

  Features:
    - Autoshadow support
    + Controls support
    + curses support
    - Developer support
    + DSO support
    + IPv6 support
    + Largefile support
    - Lastlog support
    + Memcache support
    + ncursesw support
    + NLS support
    + OpenSSL support (OpenSSL 1.1.1k  FIPS 25 Mar 2021, FIPS enabled)
    - PCRE support
    + PCRE2 support
    + POSIX ACL support
    - Redis support
    + Sendfile support
    + Shadow file support
    + Sodium support
    + Trace support
    + xattr support

  Tunable Options:
    PR_TUNABLE_BUFFER_SIZE = 1024
    PR_TUNABLE_DEFAULT_RCVBUFSZ = 8192
    PR_TUNABLE_DEFAULT_SNDBUFSZ = 8192
    PR_TUNABLE_ENV_MAX = 2048
    PR_TUNABLE_GLOBBING_MAX_MATCHES = 100000
    PR_TUNABLE_GLOBBING_MAX_RECURSION = 8
    PR_TUNABLE_HASH_TABLE_SIZE = 40
    PR_TUNABLE_LOGIN_MAX = 256
    PR_TUNABLE_NEW_POOL_SIZE = 512
    PR_TUNABLE_PATH_MAX = 4096
    PR_TUNABLE_SCOREBOARD_BUFFER_SIZE = 80
    PR_TUNABLE_SCOREBOARD_SCRUB_TIMER = 30
    PR_TUNABLE_SELECT_TIMEOUT = 30
    PR_TUNABLE_TIMEOUTIDENT = 10
    PR_TUNABLE_TIMEOUTIDLE = 600
    PR_TUNABLE_TIMEOUTLINGER = 10
    PR_TUNABLE_TIMEOUTLOGIN = 300
    PR_TUNABLE_TIMEOUTNOXFER = 300
    PR_TUNABLE_TIMEOUTSTALLED = 3600
    PR_TUNABLE_XFER_SCOREBOARD_UPDATES = 10
```

**Comment by Castaglia** [comment]
From the backtrace, I suspect that this is fixed (in master) via https://github.com/proftpd/proftpd/commit/b1ce87448b45e4efd524cba2e4f23098aa52f797.  I'll now backport that commit to the 1.3.8 branch.

**Comment by Castaglia** [comment]
Done; see https://github.com/proftpd/proftpd/commit/85ff1cb9fb2feaae9928b50d05afd657fec3cfb4.  Would you be able to verify that that addresses the behavior you're seeing?

**Comment by conathan** [comment]
I certainly can, and have.  Patched our own local version of proftpd, and no longer crashing.  Thank you.

**Comment by Castaglia** [comment]
I've now updated the [NEWS](./NEWS) file for this issue, and backported the fix to the 1.3.8 branch.
